File: python/3068.find-the-maximum-sum-of-node-values.py
class Solution:
    def maximumValueSum(self, nums: list[int], k: int, edges: list[list[int]]) -> int:
        INF = int(1e9)
        inc = 0
        inc_cnt, dec_cnt = 0, 0
        inc_min, dec_min = INF, INF
        total = 0

        for num in nums:
            total += num

            if num ^ k > num:
                inc += (num ^ k) - num
                inc_cnt += 1
                inc_min = min(inc_min, (num ^ k) - num)
            else:
                dec_cnt += 1
                dec_min = min(dec_min, num - (num ^ k))

        if inc_cnt == 0:
            return total

        if inc_cnt % 2 == 0:
            return total + inc

        if dec_cnt == 0:
            return total + inc - inc_min

        return total + inc - min(inc_min, dec_min)

File: cpp/3068.find-the-maximum-sum-of-node-values.cpp
long long min(long long a, long long b) { return a < b ? a : b; }

class Solution {
public:
  long long maximumValueSum(vector<int> &nums, int k,
                            vector<vector<int>> &edges) {
    // XOR can be applied to any pair of nodes, even without an edge connecting
    // them. since there is a path between these two nodes and there is no
    // cycles.
    long long base = 0;
    long long totalIncreaseDelta = 0;
    long long numIncrease = 0, numDecrease = 0;
    long long minIncrease = numeric_limits<long long>::max();
    long long minDecrease = numeric_limits<long long>::max();

    for (auto num : nums) {
      base += num;
      if ((num ^ k) > num) {
        numIncrease++;
        totalIncreaseDelta += (num ^ k) - num;
        minIncrease = min(minIncrease, (num ^ k) - num);
      } else {
        numDecrease++;
        minDecrease = min(minDecrease, num - (num ^ k));
      }
    }

    if (!numIncrease) {
      return base;
    }

    if (numIncrease % 2 == 0) {
      return base + totalIncreaseDelta;
    }

    if (numDecrease == 0) {
      return base + totalIncreaseDelta - minIncrease;
    }

    return base + totalIncreaseDelta - min(minIncrease, minDecrease);
  }
};